File: enrollSorter.cpp

//<!*---------- MUST BE RUN WITH C++ 11 COMPILER ----------*!>

#include <iostream>             // for input/output
#include <string>               // for strings
#include <fstream>              // for file streams
#include <vector>               // for vectors
#include <bits/stdc++.h>        // for sort() implementation

using namespace std;

struct enrollee
{
    string userid = "";
    string fname = "";
    string lname = "";
    string insurance = "";
    string version = "";
};

// INPUT: 2 enrollee structs/objects
// OUTPUT: bool value (true/false)
// FUNCTION: compare all of the values of 2 enrollee structs/objects
//              if they are NOT the same, return true
//              if they are the same, return false
//              (kind of backwards, but easier to think about when implementing the function in the code below)
bool notSameEnrollee (enrollee e1, enrollee e2)
{
    // I know I could put all this on one return line, but I think this way is more readable

    // if they have the same attributes, then they must be the same enrollee, so return false
    if (e1.userid == e2.userid && e1.fname == e2.fname && e1.lname == e2.lname && e1.insurance == e2.insurance && e1.version == e2.version)
    {
        return false;
    }
    // otherwise if they have different attributes, then they are different enrollees, so return true
    else
    {
        return true;
    }
}

int main()
{
    ifstream inputFile;               // create input file stream for reading only
    vector <enrollee> enrollVector;   // vector of enrollee structs to store each enrollee and their respective data
    int numDuplicates = 0;            // count number of enrollees received from input file w/ duplicate user id's and insurances
    string tempInsurance;             // temp var used to compare insurance strings to check for a new unique insurance

    // open the input file to read
    inputFile.open("input.csv");
    // read the file until we reach the end of the file
    while(!inputFile.eof())
    {
        enrollee tempEnrollee;
        string userid = "";
        string fname = "";
        string lname = "";
        string insurance = "";
        string sversion = "";

        // read in the cols of each row in the excel file, and store them in their respective temp vars
        getline(inputFile,userid,',');
        getline(inputFile,fname,',');
        getline(inputFile,lname,',');
        getline(inputFile,insurance,',');
        getline(inputFile,sversion);

        // assign those temp vars to an enrollee object
        tempEnrollee.userid = userid;
        tempEnrollee.fname = fname;
        tempEnrollee.lname = lname;
        tempEnrollee.insurance = insurance;
        tempEnrollee.version = sversion;

        // push the enrollee object onto the enrollVector
        enrollVector.push_back(tempEnrollee);
    }

    // close inputFile stream
    inputFile.close();

    // this call to sort will sort the enrollVector by insurance, then lname, then fname, then userid, then version
    sort( enrollVector.begin(), enrollVector.end(), []( const enrollee &e1, const enrollee &e2 )
         {
             return tie( e1.insurance, e1.lname, e1.fname, e1.userid, e1.version ) < tie( e2.insurance, e2.lname, e2.fname, e2.userid, e2.version );
         });

    // open an output file stream to write deleted enrollees to
    ofstream outputFile;
    outputFile.open("deleted.csv");

    // loop thru the enrollVector to check for duplicate id's w/ the same insurance
    for (int i = 0; i < enrollVector.size(); i++)
    {
        for (int j = 0; j < enrollVector.size(); j++)
        {
            // if the user id's AND the insurances match AND they are not the same object, then erase the one w/ the lower version
            if ( (enrollVector[i].userid == enrollVector[j].userid) && (enrollVector[i].insurance == enrollVector[j].insurance) && (notSameEnrollee(enrollVector[i], enrollVector[j])) )
            {
                // if i < j, then delete i
                if (enrollVector[i].version < enrollVector[j].version)
                {
                    // print out a message stating which enrollees were received from the input file, but deleted due to a duplicate userid/insurance
                    outputFile << enrollVector[i].userid << "," << enrollVector[i].fname << "," << enrollVector[i].lname << "," << enrollVector[i].insurance << "," << enrollVector[i].version << endl;
                    enrollVector.erase(enrollVector.begin()+i);
                    numDuplicates++;
                }
                // otherwise delete j
                else
                {
                    // print out a message stating which enrollees were received from the input file, but deleted due to a duplicate userid/insurance
                    outputFile << enrollVector[j].userid << "," << enrollVector[j].fname << "," << enrollVector[j].lname << "," << enrollVector[j].insurance << "," << enrollVector[j].version << endl;
                    enrollVector.erase(enrollVector.begin()+j);
                    numDuplicates++;
                }
            }
        }
    }
    // close the output file stream
    outputFile.close();

    // if there were duplicates that were deleted, alert the user w/ further related info
    if (numDuplicates > 0)
    {
        cout << numDuplicates << " enrollee(s) were received from the input file, but deleted due to duplicate id/insurance." << endl;
        cout << "Check deleted.csv file for list of deleted enrollees." << endl;
    }
    // otherwise just alert the user of no duplicates and delete the output file since it is empty and not needed
    else
    {
        remove("deleted.csv");
        cout << "No duplicates detected/deleted." << endl;
    }

    // write sorted data to their respective files
    for (int i = 0; i < enrollVector.size(); i++)
    {
        // if we come across a new insurance name, then start a new file for it
        if (tempInsurance != enrollVector[i].insurance)
        {
            // to reuse outputFile stream, we must close one stream before reopening another
            outputFile.close();
            tempInsurance = enrollVector[i].insurance;
            outputFile.open( tempInsurance + ".csv");
        }

        // write data to the file
        outputFile << enrollVector[i].userid << "," << enrollVector[i].fname << ","
                    << enrollVector[i].lname << "," << enrollVector[i].insurance << ","
                    << enrollVector[i].version << endl;
    }

    // close output file stream
    outputFile.close();

    return 0;
}

File: EnrollSorterJava/EnrollSorter.java
import java.io.*;
import java.util.*;

public class EnrollSorter {

	// INPUT: 2 enrollee structs/objects
	// OUTPUT: bool value (true/false)
	// FUNCTION: compare all of the values of 2 enrollee structs/objects
	//	         	if they are NOT the same, return true
	//	            if they are the same, return false
	//	            (kind of backwards, but easier to think about when implementing the function in the code below)
	public static boolean notSameEnrollee(Enrollee e1, Enrollee e2)
	{
		// if they have the same attributes, then they must be the same enrollee, so return false
		if ( (e1.getUserid().equals(e2.getUserid())) && (e1.getFname().equals(e2.getFname())) && (e1.getLname().equals(e2.getLname())) && (e1.getInsurance().equals(e2.getInsurance())) && (e1.getVersion().equals(e2.getVersion())) )
		{
			return false;
		}
		// otherwise if they have different attributes, then they are different enrollees, so return true
		else
		{
			return true;
		}
	}
	
	public static void main(String[] args) {
		
		ArrayList<Enrollee> enrolleeAL = new ArrayList<Enrollee>();
		int numDuplicates = 0;
		BufferedReader br = null;
		String inputFileName = "input.csv";
		
		// try statement for opening the input file
		try
		{
			// open the input file
			br = new BufferedReader(new FileReader(inputFileName));
			
			// print info
			System.out.println("----------------------------------------");
			System.out.println("Before Sorting");
			System.out.println("----------------------------------------");
			
			// read in the next line of the input file
			String line;
			while ((line = br.readLine()) != null)
			{
				// split up the line that was just read in, separated by commas, and store them in an array of strings
				String[] splitString = line.split(",");
				
				// create a new Enrollee object
				Enrollee tempEnrollee = new Enrollee();
				
				// assign the input values that were just read in to the new Enrollee object
				tempEnrollee.setUserid(splitString[0]);
				tempEnrollee.setFname(splitString[1]);
				tempEnrollee.setLname(splitString[2]);
				tempEnrollee.setInsurance(splitString[3]);
				tempEnrollee.setVersion(splitString[4]);
				
				// add the Enrollee object to the array list of Enrollees
				enrolleeAL.add(tempEnrollee);
			}
			
			// print enrollees before sorting
			System.out.println(enrolleeAL);
			
			// now that we have read and stored the enrollees in an array list, we must sort them by insurance, then lname, then fname, then userid, then version
			
			// sort the enrolleeAL based on the above criteria written into the class's compareTo() function
			Collections.sort(enrolleeAL);
			
			// print enrollees after sorting
			System.out.println("----------------------------------------");
			System.out.println("After Sorting");
			System.out.println("----------------------------------------");
			System.out.println(enrolleeAL);
			System.out.println("----------------------------------------\n");
			
			// open an output file for the potential deleted enrollees
			String deleteFileName = "deleted.csv";
			PrintWriter writer = new PrintWriter(deleteFileName, "UTF-8");
			
			// now that we have our sorted array list of enrollees, we must delete enrollees w/ duplicate userid's and insurances (store and write deleted entries to output file)
			
			// loop thru the enrolleeAL and compare each object to every other object to check for duplicate id's w/ the same insurance
			for (int i = 0; i < enrolleeAL.size(); i++)
			{
				for (int j = 0; j < enrolleeAL.size(); j++)
				{
					// if the user id's AND the insurances match AND they are not the same object, then erase the one w/ the lower version
					if ( (enrolleeAL.get(i).getUserid().equals(enrolleeAL.get(j).getUserid())) && (enrolleeAL.get(i).getInsurance().equals(enrolleeAL.get(j).getInsurance())) && (notSameEnrollee(enrolleeAL.get(i), enrolleeAL.get(j))) )
					{
						// if i < j, then delete i
						if ( Integer.parseInt(enrolleeAL.get(i).getVersion()) < Integer.parseInt(enrolleeAL.get(j).getVersion()) )
						{
							// print out which enrollees were received from the input file, but deleted due to a duplicate userid/insurance
							writer.println(enrolleeAL.get(i).getUserid() + "," + enrolleeAL.get(i).getFname() + "," + enrolleeAL.get(i).getLname() + "," + enrolleeAL.get(i).getInsurance() + "," + enrolleeAL.get(i).getVersion());
							
							// delete the enrollee
							enrolleeAL.remove(i);
							
							// increment duplicates counter for info later
							numDuplicates++;
							
						}
						else
						{
							// print out which enrollees were received from the input file, but deleted due to a duplicate userid/insurance
							writer.println(enrolleeAL.get(j).getUserid() + "," + enrolleeAL.get(j).getFname() + "," + enrolleeAL.get(j).getLname() + "," + enrolleeAL.get(j).getInsurance() + "," + enrolleeAL.get(j).getVersion());
							
							// delete the enrollee
							enrolleeAL.remove(j);
							
							// increment duplicates counter for info later
							numDuplicates++;
						}
					}
				}
			}
			
			// close the deleted.csv output file
			writer.close();
			
			// if there were duplicates that were detected, then alert the user w/ further related info
			if (numDuplicates > 0)
			{
				System.out.println(numDuplicates + " enrollee(s) were received from the input file, but deleted due to duplicate userid/insurance.");
				System.out.println("Check deleted.csv for list off deleted enrollee(s).\n");
				System.out.println("----------------------------------------");
			}
			// otherwise just alert the user of no duplicates and delete the output deleted.csv file since it is empty and not needed
			else
			{
				// try statement for deleting the output deleted.csv file
				try
				{
					File file = new File(deleteFileName);
					if(file.delete())
					{
						System.out.println("No duplicates detected/deleted, so deleted.csv was deleted.\n");
						System.out.println("----------------------------------------");
					}
					else
					{
						System.out.println("No duplicates detected/deleted, but there was an ERROR deleting deleted.csv.\n");
						System.out.println("----------------------------------------");
					}
				}
				catch (Exception e)
				{
					e.printStackTrace();
				}
			}
			
			// now that we are sorted and have deleted our duplicates, we just need to write them to their respective output files by insurance
			
			// open output file
			String tempInsurance = enrolleeAL.get(0).getInsurance();
			writer = new PrintWriter((tempInsurance + ".csv"), "UTF-8");
			
			// loop thru enrolleeAL and write each object to its respective insurance output file
			for (int i = 0; i < enrolleeAL.size(); i++)
			{
				// if we come across a new insurance name, then open up a new output file for that new insurance
				if ( !(tempInsurance.equals(enrolleeAL.get(i).getInsurance())) )
				{
					// close the previous output file
					writer.close();
					
					// open a new output file with the new insurance name
					tempInsurance = enrolleeAL.get(i).getInsurance();
					writer = new PrintWriter((tempInsurance + ".csv"), "UTF-8");
				}
				
				// write enrollee object to the file
				writer.println(enrolleeAL.get(i).getUserid() + "," + enrolleeAL.get(i).getFname() + "," + enrolleeAL.get(i).getLname() + "," + enrolleeAL.get(i).getInsurance() + "," + enrolleeAL.get(i).getVersion());
			}
			
			// close the output file
			writer.close();
			
		}
		// if the file was not found
		catch (FileNotFoundException e)
		{
			System.out.println("Unable to open file '" + inputFileName + "'");
		}
		// if there was an i/o error
		catch (IOException e)
		{
			e.printStackTrace();
		}
		// this will always happen once at the end no matter what
		finally
		{
			// close the file
			try {
				br.close();
			} catch (IOException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		}

	}

}
